src/flag_gems/ops/batch_norm.py

```diff
@@ -0,0 +1,489 @@
+    # and normalization would be conducted in one step.
+    # However, for large inputs, that is not feasible given memory constraints.
+    # Thus, a maximum of 16384 elements are loaded at once.
+    BLOCK_SIZE_BATCH = next_power_of_2(args["batch_dim"])
+    BLOCK_SIZE_SPATIAL = next_power_of_2(args["spatial_dim"])
+    return min(BLOCK_SIZE_SPATIAL, max(1, 2**14 // BLOCK_SIZE_BATCH))
+
+
+# NOTE: This part of the kernel code is copied and modified
+# from the https://github.com/BobMcDear/attorch codebase.
+
+
+@libentry()
+@triton.autotune(
+    configs=runtime.get_triton_config("batch_norm"),
+    key=["batch_dim", "spatial_dim"],
+    restore_value=["running_mean_pointer", "running_var_pointer"],
+)
+@triton.heuristics(
+    {
+        "BLOCK_SIZE_BATCH": lambda args: next_power_of_2(args["batch_dim"]),
+        "BLOCK_SIZE_SPATIAL": BLOCK_SIZE_SPATIAL_heuristic,
+    }
+)
+@triton.jit
+def batch_norm_forward_kernel(
+    input_pointer,
+    weight_pointer,
+    bias_pointer,
+    mean_pointer,
+    inv_std_pointer,
+    output_pointer,
+    running_mean_pointer,
+    running_var_pointer,
+    batch_dim,
+    spatial_dim,
+    input_batch_stride,
+    input_feat_stride,
+    input_spatial_stride,
+    output_batch_stride,
+    output_feat_stride,
+    output_spatial_stride,
+    momentum,
+    eps,
+    affine: tl.constexpr,
+    save_stats: tl.constexpr,
+    is_train: tl.constexpr,
+    BLOCK_SIZE_BATCH: tl.constexpr,
+    BLOCK_SIZE_SPATIAL: tl.constexpr,
+):
+    feat_pid = tl.program_id(axis=0)
+
+    batch_offset = tl.arange(0, BLOCK_SIZE_BATCH)
+    batch_mask = batch_offset < batch_dim
+
+    # traning mode default track_running_stat
+    if is_train:
+        count = 0
+        mean = 0.0
+        var = 0.0
+
+        for block_ind in range(0, tl.cdiv(spatial_dim, BLOCK_SIZE_SPATIAL)):
+            spatial_offset = block_ind * BLOCK_SIZE_SPATIAL + tl.arange(
+                0, BLOCK_SIZE_SPATIAL
+            )
+            spatial_mask = spatial_offset < spatial_dim
+
+            curr_input_pointer = (
+                input_pointer
+                + input_feat_stride * feat_pid
+                + input_batch_stride * batch_offset[:, None]
+                + input_spatial_stride * spatial_offset[None, :]
+            )
+            curr_input = tl.load(
+                curr_input_pointer, mask=batch_mask[:, None] & spatial_mask[None, :]
+            ).to(tl.float32)
+
+            spatial_count = min(
+                BLOCK_SIZE_SPATIAL, spatial_dim - block_ind * BLOCK_SIZE_SPATIAL
+            )
+            curr_count = spatial_count * batch_dim
+            count += curr_count
+
+            prev_mean = mean
+            mean += (tl.sum(curr_input) - curr_count * mean) / count
+            deltas = tl.where(
+                batch_mask[:, None] & spatial_mask[None, :],
+                (curr_input - mean) * (curr_input - prev_mean),
+                0.0,
+            )
+            var += tl.sum(deltas)
```

Comment:
this method cannot fully utilize vectorization/tensorization and leads to sequential computation.